### ISSUE TITLE
Disallow proc / func keywords in OSH to avoid misleading syntax errors

### ISSUE THREAD

**Issue by bar-g** [oilshell/oil#1762]
Stumped by this one, given all the examples on the website, and especially with the error not giving any hint.

Minimal example:

```
osh-0.19.0$ proc foo () { true; }

  proc foo () { true; }
           ^
[ interactive ]:34: Empty arg list not allowed
```

**Comment by andychu** [comment]
Oh sorry I just ran into this too - https://oilshell.zulipchat.com/#narrow/stream/384942-language-design/topic/OSH.20can.20give.20bad.20errors.20on.20YSH.20code

The problem is that you need to use YSH

This error message is very confusing because in OSH `proc` is not a keyword, so it's parsing it as a CALL rather than a definition

So `ls foo ()` is illegal

I think the answer is to reserve `proc` and `func` in OSH too, so we can give better error messages

**Comment by bar-g** [comment]
Hi, there! Yeah, thought about it not just to allow better error messages, but also to allow easy migration of selective functions into a proc (implicit ysh settings?), e.g. to try out some feature like simply placing some output @var instead of a dance with `read -r -d ''` and related pitfalls.

**Comment by andychu** [comment]
Yeah I think there was a discussion about proc in OSH awhile ago

The decision was to only allow proc and func with `shopt --set ysh:upgrade`

The reason being is that I don't want two semantics for a proc -- errexit and no errexit.   It should always be errexit

And the problem with mixing is that you get **mixed stacks** -- proc calling shell function calling proc calling shell function

This is too confusing

So basically proc is for YSH only.  If you want to use them, then you have to do `shopt --set ysh:upgrade`

This could probably be a design FAQ

**Comment by bar-g** [comment]
Thanks, very reasonable update!

> parsing it as a CALL rather than a definition

Hm, is osh interpreting the parenthesis as a subshell, then? ( I tried `echo ()`)  Then, maybe making the error message reflect that might be enough of a hint to show (or know) that proc needs an explicit `shopt --set ysh:upgrade`.

**Comment by bar-g** [comment]
The shell prompt that I got does not seem to be specific to `echo ( )`, but but seems to be a different peculiarity, it  occurs with `ls ( )` as well, maybe depending if it is a first or second argument.

A naive question would be if the first word of a call could be resolved before (or in parallel to) the argument list of a call, if the error message were to be something like `'proc' not found`, that might also make it obvious ysh would need to be enabled.

**Comment by andychu** [comment]
Yes good point about `echo ()`, I filed a separate bug #1767 for this

(The reason is that echo uses a special flag parser due to lack of `--`, and this accidentally made it accept `(x)`)

**Comment by andychu** [comment]
Fixed, will be out with the next release

**Comment by andychu** [comment]
Fixed - https://www.oilshell.org/blog/2024/03/release-0.21.0.html
